File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(num)
    raise RuntimeError if num < 0
    factor_sum = PerfectNumber.parse_factors(num).reduce(:+)
    factor_sum > num ? 'abundant'
                     : (factor_sum < num ? 'deficient'
                                         : 'perfect'
                       )
  end

  private

  def self.parse_factors(num)
    sum = [1]
    (2..Math.sqrt(num)).select { |e| num % e == 0 }
               .each { |e| sum << e; sum << num/e }
    sum.uniq
  end
end

module BookKeeping
  VERSION = 1
end

File: python/perfect-numbers/perfect_numbers.py
import math

def classify(number):
    if number < 1:
        raise ValueError('invalid number')
    factors = parse_factors(number)

    if sum(factors) == number and number !=  1:
        return 'perfect'
    elif sum(factors) > number:
        return 'abundant'
    else:
        return 'deficient'


def parse_factors(number):
    factors = [1]
    for i in range(2, math.ceil(math.sqrt(number))):
        if number % i == 0:
            factors.append(i)
            factors.append(number // i)
    return factors